### ISSUE TITLE
Consider to implicitly apply Serializable annotation to the entire sealed hierarchy

### ISSUE THREAD

**Issue by arkivanov** [Kotlin/kotlinx.serialization#2572]
**What is your use-case and why do you need this feature?**

Parcelize allows the following code.

```kotlin
@Parcelize
sealed interface State : Parcelable {
    data object Loading : State
    data object Error : State
    data class Data(val data: String) : State
}
```

We are able to specify the annotation only once on the base `State` interface. With `kotlinx-serialization` we have to repeat the annotation for every implementing class, which is pretty verbose.

```kotlin
@Serializable
sealed interface State {
    @Serializable
    data object Loading : State

    @Serializable
    data object Error : State

    @Serializable
    data class Data(val data: String) : State
}
```

**Describe the solution you'd like**

It would be nice to be able to write as follows.

```kotlin
@Serializable
sealed interface State {
    data object Loading : State
    data object Error : State
    data class Data(val data: String) : State
}
```

Alternative API if changing the existing behaviour is undesirable.

```kotlin
@Serializable(inherited = true)
sealed interface State {
    data object Loading : State
    data object Error : State
    data class Data(val data: String) : State
}
```

**Comment by sandwwraith** [comment]
Given that inheritors of the sealed interface can be defined anywhere in the project, and it is not mandatory for them to be in the same file, this would be hard from a technical standpoint: it would require the plugin to analyze every single class in the project and its superclasses, instead of focusing on `@Serializable` classes. It also brings questions for design: if my intention to have only some of the inheritors serializable, how can I opt-out?

To sum up, we cannot change the behavior of the existing `@Serializable` annotation, as it would not be backward-compatible. It is possible to implement additional annotation though, that would cover the exact use case you've presented here. E.g. `@SerializableSubclasses` for nested classes or classes in the same file.

**Comment by arkivanov** [comment]
`@SerializableSubclasses` sounds good as well. What about adding an optional argument to `@Serializable`? E.g. `inherited: Boolean` with default value `false` to keep the current behaviour as default, and make the new feature opt-in.

**Comment by sandwwraith** [comment]
`@Serializable` also can be applied to the properties to specify custom serializer, so `inherited` parameter there would look illogical.

**Comment by sandwwraith** [comment]
Duplicate of #2640, I guess

**Comment by arkivanov** [comment]
Indeed, though this issue was created earlier, I don't mind closing it in favour of another one.
